zookeeper-server/src/main/java/org/apache/zookeeper/server/DataTree.java

```diff
@@ -808,74 +840,163 @@ public int compare(PathWithStat left, PathWithStat right) {
             }
-            return result;
         }
+
+        updateNextPage(outputNextPage, targetChildren, index);
+        updateReadStat(path, countReadChildrenBytes(paginatedChildren));
+
+        return paginatedChildren;
+    }
+
+    private boolean canPacketFitInMaxBuffer(int packetLength) {
+        return packetLength <= BinaryInputArchive.maxBuffer;
+    }
+
+    private int countReadChildrenBytes(Collection<String> children) {
+        return children.stream().mapToInt(String::length).sum();
+    }
+
+    private void buildChildrenPathWithStat(DataNode n, String path, Stat stat, long minCzxId,
+                                           List<PathWithStat> targetChildren) {
+        synchronized (n) {
+            if (stat != null) {
+                n.copyStat(stat);
+            }
+            for (String child : n.getChildren()) {
+                DataNode childNode = nodes.get(path + "/" + child);
+                if (null != childNode) {
+                    final long czxId = childNode.stat.getCzxid();
+
+                    if (czxId < minCzxId) {
+                        // Filter out nodes that are below minCzxId
+                        continue;
+                    }
+
+                    Stat childStat = new Stat();
+                    childNode.copyStat(childStat);
+
+                    // Cannot discard before having sorted and removed offset
+                    targetChildren.add(new PathWithStat(child, childStat));
+                }
+            }
+        }
+    }
+
+    /*
+     * totalLength = (PAGINATION_PACKET_CHILD_EXTRA_BYTES + childLength) * numChildren + PAGINATION_PACKET_BASE_BYTES
+     */
+    private int computeChildrenPacketLength(Collection<String> children) {
+        int length = children.stream().mapToInt(child -> PAGINATION_PACKET_CHILD_EXTRA_BYTES + child.length()).sum();
+        return length + PAGINATION_PACKET_BASE_BYTES;
+    }
+
+    private void updateNextPage(PaginationNextPage nextPage, List<PathWithStat> children, int lastAddedIndex) {
+        if (nextPage == null) {
+            return;
+        }
+        if (lastAddedIndex == children.size() - 1) {
+            // All children are added, so this is the last page
+            nextPage.setMinCzxid(ZooDefs.GetChildrenPaginated.lastPageMinCzxid);
+            return;
+        }
+
+        // Find the minCzxidOffset next next page by searching the index (startIndex) of czxid
+        // that is not equal to current czxid.
+        // minCzxidOffset of next page = lastAddedIndex - startIndex
+        long lastCzxid = children.get(lastAddedIndex).getStat().getCzxid();
+        int startIndex = lastAddedIndex;
+        while (startIndex >= 0) {
+            if (children.get(startIndex).getStat().getCzxid() != lastCzxid) {
+                break;
+            }
+            startIndex--;
+        }
+        nextPage.setMinCzxid(lastCzxid);
+        nextPage.setMinCzxidOffset(lastAddedIndex - startIndex);
```

Comment:
Could this nextPage shared by another thread? Say there are two paginated calls from same client. Are they using the same object?

Reply:
On the server side, no. Each call has its own object. The object is created in the `FinalRequestProcessor`: https://github.com/linkedin/zookeeper/blob/c400bc183a7c61f35a524a57f828715d103badf9/zookeeper-server/src/main/java/org/apache/zookeeper/server/FinalRequestProcessor.java#L579

Reply:
Set offset = 0 for next zxid.